src/parquet/bloom_filter.h

```diff
@@ -77,51 +42,209 @@ class BloomFilter final {
+
+  /// Get the number of bytes of bitset
+  virtual uint32_t GetBitsetSize() const = 0;
 
-  /// Compute hash for value by using its plain encoding result.
+  /// Compute hash for 32 bits value by using its plain encoding result.
   ///
   /// @param value the value to hash.
   /// @return hash result.
-  template <typename T>
-  uint64_t Hash(T value) const {
-    return hasher_->Hash(value);
-  }
+  virtual uint64_t Hash(int32_t value) const = 0;
+
+  /// Compute hash for 64 bits value by using its plain encoding result.
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(int64_t value) const = 0;
+
+  /// Compute hash for float value by using its plain encoding result.
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(float value) const = 0;
+
+  /// Compute hash for double value by using its plain encoding result.
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(double value) const = 0;
 
-  /// Compute hash for Fixed Length Byte Array value by using its plain encoding result.
+  /// Compute hash for Int96 value by using its plain encoding result.
   ///
   /// @param value the value to hash.
   /// @return hash result.
-  uint64_t Hash(const FLBA* value, uint32_t len) const {
+  virtual uint64_t Hash(const Int96 *value) const = 0;
+
+  /// Compute hash for ByteArray value by using its plain encoding result.
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(const ByteArray *value) const = 0;
+
+  /// Compute hash for fixed byte array value by using its plain encoding result.
+  ///
+  /// @param value the value to hash.
+  /// @return hash result.
+  virtual uint64_t Hash(const FLBA *value, uint32_t len) const = 0;
+
+  virtual ~BloomFilter() {}
+
+ protected:
+  // Hash strategy available for Bloom filter.
+  enum class HashStrategy : uint32_t {
+    MURMUR3_X64_128 = 0
+  };
+
+  // Bloom filter algorithm.
+  enum class Algorithm : uint32_t {
+    BLOCK = 0
+  };
+};
+
+// The BlockSplitBloomFilter is implemented using block-based Bloom filters from
+// Putze et al.'s "Cache-,Hash- and Space-Efficient Bloom filters". The basic idea is to
+// hash the item to a tiny Bloom filter which size fit a single cache line or smaller.
+//
+// This implementation sets 8 bits in each tiny Bloom filter. Each tiny Bloom
+// filter is 32 bytes to take advantage of 32-byte SIMD instructions.
+class BlockSplitBloomFilter : public BloomFilter {
+ public:
+  /// The constructor of BlockSplitBloomFilter. The range of num_bytes should be within
+  /// [MINIMUM_BLOOM_FILTER_BYTES, MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// rounded up/down to lower/upper bound if num_bytes is out of range and also
+  /// will be rounded up to a power of 2. It will use murmur3_x64_128 as its
+  /// default
+  /// hash function and the block-based algorithm.
```

Comment:
Don't need to say the default algorithm, only the hash function. Here and below.